accepted/single-file/design.md

```diff
@@ -0,0 +1,281 @@
+
+- R2R compile the app and all of its dependent assemblies in a single version-bubble
+- Investigate whether collectively signing the files in an assembly saves space for certificates.
+
+#### APIs to access Bundled files
+
+The binaries that are published in the project are expected to be handled transparently by the host. However, explicit access to the embedded files is useful in situations such as:
+
+- Reading additional files packaged into the app (ex: data files).
+- Open an assembly for reflection/inspection
+- Load plugins built as single-file class-libs using existing Loader APIs.
+
+In facilitate this usage, we can add an API similar to [GetManifestResourceStream](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.getmanifestresourcestream?view=netframework-4.7.2#System_Reflection_Assembly_GetManifestResourceStream_System_String_) to obtain a stream corresponding to an embedded file. 
+
+```C#
+// Open a file embedded in the bundle built for the specified assembly 
+namespace System.Runtime.Loader
+{
+    public partial class Bundle
+    {
+        public static System.IO.Stream GetFileStream(Assembly assembly, string name);
+    }
+}
+```
+
+We can also provide an abstraction that abstracts away the physical location of a file (bundle or disk). For example, add a variant of  `GetFileStream` API that looks for a file in the bundle, and if not found, falls back to disk-lookup.
+
+#### Single-file Plugins
+
+The above design should be extended to seamlessly support single-file publish for plugins.
+
+- The bundler tool will mostly work as-is, regardless of whether we publish an application or class-lib. The binary blob with dependencies can be appended to both native and managed binaries.
+- For host/runtime support, the options are:
+  - Implement plugins using existing infrastructure. For example: Take control of assembly/native binary loads via existing `AssemblyLoadContext` callbacks and events. Extract the files embedded within the single-file plugin using the `GetFileStream()` API and load them on demand.
+  - Have new API to load a single-file plugin, for example: `AssemblyLoadContext.LoadWithEmbeddedDependencies()`.
+
+#### Project File Options
+
+By default, `PublishSingleFile` bundles all files in the publish directory into the single executable. This behavior can be altered via the following configurations in the application's project file.
+
+* Exclude certain files in the publish directory from being bundled (ex: because of licensing issues or for testing purposes).
+
+```xml
+<ItemGroup>
+    <SingleFileExclude Include="List of patterns to exclude"/>
+</ItemGroup>
+```
+
+- Add additional files to the bundle from other locations (ex: add data files to the app). The developer will be responsible for handling these additional files.
+
+```xml
+<ItemGroup>
+    <SingleFileAdd Include="List of file paths"/>
```

Comment:
There are IDE considerations for adding new item types. This may not give the best experience for maintaining such files in the IDE.

We should consider if this can be metadata on Content. Today there is CopyToOutputDirectory, CopyToPublishDirectory. There can be IncludeInSingleFilePublish or something.

Today, CopyToOutputDirectory implies CopyToPublishDirectory by default and you can override.

Reply:
That's a good suggestion @nguerrera, I'll change `SingleFile Include/Exclude` to be meta-data.

Reply:
@nguerrera I've updated the doc now to use `IncludeInSingleFilePublish` metadata.